_posts/2017-04-21-calculating-the-importance-of-an-artwork-with-apache-spark.markdown

```diff
@@ -0,0 +1,141 @@
+     |  SELECT follow_artists.artist_id, COUNT(*) AS artist_follows_count FROM db.follow_artists GROUP BY follow_artists.artist_id
+     |) artist_follows_count ON artist_follows_count.artist_id = artwork.artist_id
+    """.stripMargin
+  ).rdd.map {
+    row =>
+      Artwork(
+        id = row.getString(0)
+        partnerType = row.getString(1)
+        artistFollowsCount = row.getLong(2)
+      )
+  }
+}
+```
+
+Some features require minor transforms. For example, the fact that the work belongs to a museum scores 1, and otherwise scores a 0.
+
+```scala
+case class Features(
+  val artworkId: Int,
+  val artistFollowsCount: Long = 0,
+  val isMuseumWork: Int = 0
+)
+
+def extractFeatures(artworks: RDD[Artwork]): RDD[Features] = {
+  artworks.map { artwork =>
+    Features(
+      artworkId = artwork.id,
+      artistFollowsCount = artwork.artistFollowsCount,
+      isMuseumWork = if (artwork.isMuseumWork) 1 else 0
+    )
+  }
+}
+```
+
+Features are packed in a vector to become usable by the built-in Spark functions.
+
+```scala
+case class Features(
+   ...
+) = {
+  def vector: Vector = {
+    Vectors.dense(
+      numArtistFollowers.toDouble,
+      isMuseumWork.toDouble
+    )
+  }
+}
+```
+
+### Normalizing and Weighing Features
+
+Since having 10,000 artist followers doesn't make a work 10,000 times more important than the fact that it belongs to a museum, we must normalize them across the entire data set. This is also a good time to weigh some features more than others according to our understanding of the art world.
+
+```scala
+import org.apache.spark.mllib.feature._
+import org.apache.spark.mllib.linalg._
+
+def normalize(features: RDD[Features]): RDD[(Int, Vector)] = {
+  val scaler = new StandardScaler().fit(features.map(f => f.vector))
+
+  val weightAdjuster = new ElementwiseProduct(Vectors.dense(
+    0.5, // number of users following an artist is a popularity contest
+    2.0  // having a work in a museum is a big deal
+  ))
+
+  features.map { f =>
+    (
+      f.artworkId,
+      weightAdjuster.transform(
+        scaler.transform(f.vector)
+      )
+    )
+  }
+}
+```
+
+### Scoring Artworks
+
+The score is just the sum of the normalized and weighted features.
+
+```scala
+def score(normalizedFeatures: RDD[(Int, Vector)]): RDD[(Int, Double)] = {
+  normalizedFeatures.map {
+    f => (f._1, f._2.toArray.sum)
+  }
+}
+```
+
+### Storing Data
+
+We write this data in JSON format to S3, then load it in a system that serves the Artsy API.
+
+### Conclusion
+
+In our dataset this creates a nice distribution. Here's an example of iconicity across [works by the street artist Banksy](https://www.artsy.net/artist/banksy/works).
+
+![banksy iconicity](/images/2017-04-21-calculating-the-importance-of-an-artwork-with-apache-spark/banksy-iconicity.png)
+
+We notably sort works by iconicity in [search results](https://www.artsy.net/search?q=banksy) and are have made it available in our [public API](https://developers.artsy.net).
```

Comment:
I think not since it's not really relevant to the post proper and is more of our own legacy, even though that's an impressive difference. I did mention that the whole thing takes under five minutes for the entire data set in the intro.